src/Symfony/Component/Clock/Clock.php

```diff
@@ -0,0 +1,58 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Clock;
+
+use Psr\Clock\ClockInterface;
+
+/**
+ * A global clock.
+ *
+ * Note that you should prefer injecting a ClockInterface or using
+ * ClockAwareTrait when possible instead of using this class.
+ *
+ * Defined as enum to be final and non-instantiable.
+ */
+enum Clock
+{
+    public static function now(): \DateTimeImmutable
+    {
+        return self::get()->now();
+    }
+
+    public static function with(ClockInterface $clock, callable $callback, mixed ...$arguments): mixed
+    {
+        $prevClock = self::get($clock);
+
+        try {
+            return $callback(...$arguments);
+        } finally {
+            self::get($prevClock);
+        }
+    }
+
+    /**
+     * Returns the current clock, or the previous one if a new clock is passed as argument.
+     */
+    public static function get(ClockInterface $newClock = null): ClockInterface
```

Comment:
this `get` method being also the setter looks like a confusing API to me. I suggest switching to a final class with private constructor (which still prevents instantiation) which would allow using a static class property to hold the state while allowing multiple methods to access it.

Reply:
I agree with stof. It looks confusing to me too.

Reply:
thanks for pushing this concern. Here is an update: https://github.com/symfony/symfony/pull/48642#issuecomment-1357714027